mqt/qcec/symbolic.py

```diff
@@ -0,0 +1,147 @@
+
+
+def check_instantiated(
+    circ1: QuantumCircuit, circ2: QuantumCircuit, **kwargs: Any
+) -> EquivalenceCheckingManager.Results:
+    """Check circuits for equivalence with dd equivalence checker."""
+    ecm = EquivalenceCheckingManager(circ1, circ2, **kwargs)
+    ecm.set_zx_checker(False)
+    ecm.set_alternating_checker(True)
+    ecm.set_simulation_checker(True)
+
+    ecm.run()
+
+    print(circ1.draw())
+    print(circ2.draw())
+    return ecm.get_results()
+
+
+def check_instantiated_random(
+    circ1: QuantumCircuit, circ2: QuantumCircuit, params: set(Parameter), **kwargs: Any
+) -> EquivalenceCheckingManager.Results:
+    """Check whether circuits are equivalent for random instantiation of symbolic parameters."""
+    param_map = {}
+    for p in params:
+        param_map[p] = np.random.rand() * 2 * np.pi
+
+    circ1_inst = circ1.bind_parameters(param_map)
+    circ2_inst = circ2.bind_parameters(param_map)
+    ecm = EquivalenceCheckingManager(circ1_inst, circ2_inst, **kwargs)
+    ecm.set_zx_checker(False)
+    ecm.set_alternating_checker(True)
+    ecm.set_simulation_checker(True)
+
+    ecm.run()
+
+    return ecm.get_results()
+
+
+def check_symbolic(
+    circ1: QuantumCircuit, circ2: QuantumCircuit, n_checks: int = 6, tol: float = 1e-15, **kwargs: Any
+) -> EquivalenceCheckingManager.EquivalenceCheckingManager.Results:
+    """Equivalence checking flow for parameterized circuit."""
+    res = check_symbolic_zx(circ1, circ2, **kwargs)
+
+    if res.considered_equivalent():
+        return res
+
+    parameters, A, offsets = extract_params(circ1, circ2)
+
+    def instantiate_params(i, qc1: QuantumCircuit, qc2: QuantumCircuit) -> tuple[QuantumCircuit, QuantumCircuit]:
+        phases = [0, np.pi, np.pi / 2, -np.pi / 2, np.pi / 4, -np.pi / 4]
+        if i > 0:
+            b = np.random.choice(phases, size=len(offsets)) + offsets
+        else:
+            b = offsets
+
+        A_pinv = np.linalg.pinv(A)
+        x = np.dot(A_pinv, b)
+        param_map = {param: x[i] for i, param in enumerate(parameters)}
+        qc1_bound = qc1.bind_parameters(param_map)
+        qc2_bound = qc2.bind_parameters(param_map)
+
+        def round_zero_params(qc: QuantumCircuit) -> QuantumCircuit:
+            for instr in qc.data:
+                params = instr[0].params
+                instr[0].params = [float(x) for x in params]
+                instr[0].params = [0 if np.abs(x) < tol else x for x in instr[0].params]
+            return qc
+
+        qc1_bound = round_zero_params(qc1_bound)
+        qc2_bound = round_zero_params(qc2_bound)
+
+        return (qc1_bound, qc2_bound)
+
+    for i in range(0, n_checks - 1):
+
+        circ1_inst, circ2_inst = instantiate_params(i, circ1, circ2, **kwargs)
+        res = check_instantiated(circ1_inst, circ2_inst)
+        if res.equivalence == "not_equivalent":
+            return res
+
+    res = check_instantiated_random(circ1, circ2, parameters, **kwargs)
```

Comment:
following up on the comment above, this could probably be a bit more transparent.
there is _one_ run with precisely determined RHS and _one_ last resort random instantiation.
in-between, there are the other attempts.

Reply:
Good point. For symmetries' sake I also made a dedicated method for the first instance